### ISSUE TITLE
speaker: three standalone faces for the decomposed diarization pipeline

### ISSUE THREAD

**Issue by uqio** [findit-studio/coremlit#110]
## Context

mediagraph is decomposing speaker diarization into **three autonomous compute nodes** (ruled 2026-08-30): `segmentation` → `embed` → `cluster`, running parallel to ASR with both pipelines subscribing the vad node (vad's `SpeechBatch` already carries `spans + samples: Arc<[f32]>`). Per-stage nodes give per-stage performance control (own worker pools, own inflight bounds), keep raw PCM out of the cluster stage (it accumulates embeddings only — KBs per track instead of the single-node form's ~461 MiB/track PCM retention), and put embeddings on the wire for future voiceprint / speaker-entity work.

Today's public face is whole-track and fully internal:

```rust
ModelSource::extract(&self, samples: &[f32]) -> Extraction   // whole track in; chunking/segmentation/embedding all inside
Extraction::diarize_with(&self, plda, backend) -> OfflineOutput  // embeddings die inside; only RttmSpans come out
```

mediagraph#135 (a single-node mirror of this face) is merged and serves in the interim. The three faces below are the walls the decomposition needs.

## 1 · A streaming / segment-level segmentation face

The segmentation node wants to push bounded audio in (per vad speech batch) and receive speaker-activity segments out, incrementally — not one whole-track call.

**Census question first**: inside today's `extract`, is segmentation a real model stage (pyannote-class speaker-activity detection) or a fixed sliding-window chunking policy? If the former, expose that stage standalone and incremental; if the latter, this ask becomes "introduce a segmentation stage" and should say so explicitly.

## 2 · A per-segment embedding face

`embed(samples_of_one_segment) -> Embedding` (plus a batch form for throughput), callable independently of whole-track extraction, so the embed node can process one segment message at a time.

## 3 · Embeddings as first-class values, in and out

- **Out**: raw embeddings leave the engine as values (today they exist per chunk inside `Extraction` and do not survive clustering).
- **In**: a clustering face decoupled from `Extraction`'s private state — `cluster(embeddings, plda, backend) -> OfflineOutput` — so the cluster node can accumulate a track's embeddings across messages and cluster at track end.

The out-face doubles as the raw material for voiceprint retrieval and the speaker entity layer later.

## Non-goals here

No wire/topology opinions — those live in mediagraph's design book. This issue is only the engine surface.

**Comment by uqio** [comment]
## Census answer: segmentation is a real model stage, not a chunking policy

`crates/coremlit/src/audio/speaker/segment/mod.rs` is a CoreML wrapper for **`pyannote_segmentation.mlmodelc`** — `audio [1, 1, 160_000]` f32 in, `segments [1, 589, 7]` f32 out — plus the powerset→multilabel decode. It ports the model-facing half of dia's `segment` stage.

So ask #1 resolves to your first branch: **expose the existing stage**, not "introduce a segmentation stage".

## Three of the four asks are already public

I audited the actual surface rather than the summary in the issue. Most of what's being asked for exists today:

| ask | status | what already exists |
|---|---|---|
| 1 · segmentation, standalone | **public today** | `SegmentModel::{from_file, from_file_with, infer, num_frames}`, `multilabel()`, `SEG_CHUNK_SAMPLES`, `SEG_NUM_SLOTS`, `POWERSET_CLASSES`, `SegmentModelOptions` |
| 1 · incremental driver | **missing** | see below |
| 2 · per-segment embed | **public today** | `EmbedModel::embed_chunk_with_frame_mask(samples, frame_mask) -> [f32; 256]` — literally one segment in, one embedding out. `embed_chunk` gives the 3-slot form. |
| 2 · batch form | missing | no batched entry point |
| 3 · embeddings **out** | **public today** | `Extraction::raw_embeddings() -> &[f32]`, plus `num_chunks()`, `num_speakers()`, `segmentations()`, `count()`, `chunks_sw()`, `frames_sw()`, `into_offline_input()` |
| 3 · clustering face **in** | **the real gap** | see below |

**One correction to the issue text.** "embeddings die inside; only `RttmSpans` come out" is not accurate — `Extraction::raw_embeddings()` is public, and so is every other field the clusterer consumes. The out-direction is already open. The blocker is the **in**-direction.

## The two genuine gaps

**A · the incremental segmentation driver.** `SegmentModel::infer` is deliberately *not* incremental: it demands exactly `SEG_CHUNK_SAMPLES` (160 000) and **rejects** short input with `InferError::InputLength` rather than padding. That is intentional and documented — it mirrors dia's contract, where zero-padding happens one layer up in `Segmenter::emit_window` / `OwnedDiarizationPipeline`'s chunk loop, and coremlit turned dia's `debug_assert_eq!` into a checked `Result`.

The sliding-window loop, the zero-padding of the final short chunk, and the stitching all live inside `Extractor::extract`. A segmentation node today would have to reimplement that loop — and getting the padding or the window arithmetic subtly wrong is exactly the class of bug that does not fail loudly. **That driver is what ask #1 actually needs**, not the model wrapper.

**B · `Extraction` cannot be built from outside.** `Extraction::from_parts` is `pub(crate)`, and `diarize_with` / `diarize_online` are methods *on* `Extraction`. So there is no path for a cluster node to accumulate embeddings across messages and cluster at track end — it has the accessors to read an `Extraction` apart, and no way to put one back together.

This is the smallest, highest-leverage change of the three: `cluster(embeddings, plda, backend) -> OfflineOutput` as you spec'd it needs either a public constructor for `Extraction` or a free function taking the parts. Everything downstream of it already works.

## One caveat on the out-face for voiceprints

`raw_embeddings()` is laid out `[chunk][slot][dim]` — per sliding-window chunk and speaker slot, with dropped `(chunk, slot)` rows all-zero. That is the right shape for clustering, which is what it was built for. It is **not** per speech segment, so a voiceprint / speaker-entity layer consuming it would also need `count()` and `segmentations()` to resolve which chunk-slots belong to one speaker turn. Worth deciding early whether the out-face should be the raw chunk×slot tensor or a per-segment view, because that choice propagates into the wire format.

## Suggested sequencing

1. **B first** — smallest surface, unblocks the cluster node, and needs no new model work.
2. **A next** — extract the chunk loop out of `Extractor` into a reusable incremental driver, so segmentation-node and `extract` share one implementation rather than two.
3. **#2's batch form** last — it is a throughput optimisation over an API that already works.

Nothing here needs a new model or a new CoreML artifact.

**Comment by uqio** [comment]
## Starting with ask #3's in-face — and its proposed signature cannot work as written

`cluster(embeddings, plda, backend) -> OfflineOutput` is under-specified. Clustering does not consume embeddings alone. `diaric::offline::OfflineInput` needs **seven independent data inputs** (plus tuning), and they come from **two different stages**:

| input | produced by |
|---|---|
| `raw_embeddings: &[f32]` | the **embed** stage |
| `segmentations: &[f64]` | the **segmentation** stage |
| `count: &[u8]` | segmentation stitching (per-output-frame speaker count) |
| `num_chunks`, `num_frames_per_chunk` | window geometry |
| `chunks_sw`, `frames_sw` | window geometry |

(`num_speakers` is not an input — it is the fixed `SEG_NUM_SLOTS` = 3. `num_output_frames` is not either — it *is* `count.len()`, derived so the two cannot disagree.)

**Consequence for the topology.** The issue describes `segmentation → embed → cluster` as a chain, but the cluster node needs segmentation's outputs as well as embed's. So either the cluster node is a **fan-in** from both upstream stages, or the embed node **forwards** the segmentation payload through as part of its message. That is a wire-format decision and it belongs in mediagraph's design book, not here — but it has to be made before the cluster node can be built, and the chain as drawn does not supply it.

I am flagging it rather than choosing, since the issue explicitly scopes wire/topology out.

## What coremlit is getting

The engine-side answer is the same under either topology: **make `Extraction` publicly constructible**, since it already is exactly the carrier for those seven inputs and already owns every method the cluster node wants — `diarize_with`, `diarize_online`, `into_offline_input`.

Today `Extraction::from_parts` is `pub(crate)`, so an external node can take an `Extraction` apart through the public accessors and has no way to put one back together. That asymmetry is the whole blocker; it is a smaller change than a new free function, and it avoids a second surface that would have to be kept in step with the first.

The public constructor will **validate**, which the `pub(crate)` one does not — it trusts its in-crate callers. The invariants are already documented on the accessors:

- `raw_embeddings.len() == num_chunks * num_speakers * EMBEDDING_DIM`
- `segmentations.len() == num_chunks * num_frames_per_chunk * num_speakers`
- `num_output_frames == count.len()`

Handing an unvalidated constructor to a node that assembles parts from two upstream stages across many messages would turn a dropped or misordered message into garbage clusters or a panic inside `diaric`, far from the cause. So the parts arrive as a named struct rather than seven positional arguments — four of them are `usize`, and transposing `num_chunks` with `num_frames_per_chunk` would compile.

Then the issue's requested shape is just:

```rust
Extraction::try_from_parts(parts)?.diarize_with(&plda, backend)
```

Asks #1's incremental driver and #2's batch form follow after this lands.

**Comment by uqio** [comment]
Closing: all three faces are on `main`.

- **Ask 1 — segment-level segmentation face:** `audio::speaker::segment::SegmentModel::infer` (one `SEG_CHUNK_SAMPLES` chunk in, `[1, num_frames, POWERSET_CLASSES]` logits out, validated per call) plus `segment::multilabel` for the powerset→multilabel decode. The chunk is the unit, so a node drives it incrementally.
- **Ask 2 — per-segment embedding face:** `audio::speaker::extract::Extractor::extract_chunk_embeddings(embed, chunk, &mut slab)` — the embed node's work with no segmenter in sight; its doc example is written as the segmentation→embed→cluster split.
- **Ask 3 — embeddings as first-class values, in and out:** `Extraction::try_from_parts(ExtractionParts)` (validating; PR #111) on the way in, the `raw_embeddings` / `segmentations` / `count` accessors and `into_offline_input` on the way out; `diarize_with` / `diarize_online` unchanged.

The wire/topology question raised in the second comment (cluster is a fan-in from both upstream stages, or embed forwards segmentation's payload) stays mediagraph's to decide; nothing in coremlit depends on the answer.
